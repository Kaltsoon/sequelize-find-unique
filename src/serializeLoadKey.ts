import { WhereOptions } from 'sequelize';
import jsonStringify from 'json-stable-stringify';

import { isPrimitive } from './utils';
import { FindUniqueOptions } from './types';

const parseWhere = (options: FindUniqueOptions<any>): WhereOptions => {
  if (!options.where || typeof options.where !== 'object') {
    throw new Error('Where parameter must be an object');
  }

  const { where } = options;

  const entries = Object.entries(where);

  for (const [key, value] of entries) {
    if (!isPrimitive(value)) {
      throw new Error(
        `Where parameter properties can only contain serializable primitive values. Found value of type "${typeof value}" in property "${key}"`,
      );
    }
  }

  return where;
};

const serializeLoadKey = (options: FindUniqueOptions<any>): string => {
  const where = parseWhere(options);

  return jsonStringify(where);
};

export default serializeLoadKey;
