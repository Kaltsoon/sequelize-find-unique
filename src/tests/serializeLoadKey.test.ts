import serializeLoadKey from '../serializeLoadKey';

describe('serializeLoadKey', () => {
  test('serializes primitive values correctly', () => {
    expect(
      serializeLoadKey({
        where: {
          firstName: 'Kalle',
          lastName: 'Ilves',
        },
      }),
    ).toMatchSnapshot();
  });

  test('throws error for non-primitive values', () => {
    expect(() =>
      serializeLoadKey({
        where: {
          firstName: 'Kalle',
          lastName: 'Ilves',
          friends: ['elina'],
        },
      }),
    ).toThrowErrorMatchingSnapshot();
  });

  test('ignores non-where parameters', () => {
    expect(
      serializeLoadKey({
        where: {
          firstName: 'Kalle',
          lastName: 'Ilves',
        },
        attributes: ['firstName', 'lastName'],
      }),
    ).toMatchSnapshot();
  });
});
