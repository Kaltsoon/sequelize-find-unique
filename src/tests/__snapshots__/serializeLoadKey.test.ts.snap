// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeLoadKey ignores non-where parameters 1`] = `"{\\"firstName\\":\\"Kalle\\",\\"lastName\\":\\"Ilves\\"}"`;

exports[`serializeLoadKey serializes primitive values correctly 1`] = `"{\\"firstName\\":\\"Kalle\\",\\"lastName\\":\\"Ilves\\"}"`;

exports[`serializeLoadKey throws error for non-primitive values 1`] = `"Where parameter properties can only contain serializable primitive values. Found value of type \\"object\\" in property \\"friends\\""`;
